lib/hal_client/link.rb

```diff
@@ -0,0 +1,108 @@
+require 'hal_client/representation'
+
+class HalClient
+
+  # HAL representation of a single link. Provides access to an embedded representation.
+  class Link
+
+    # Create a new Link
+    #
+    # options - name parameters
+    #   :rel - This Link's rel property
+    #   :target - An instance of Representation
+    #   :template = A URI template ( https://www.rfc-editor.org/rfc/rfc6570.txt )
+    def initialize(options)
+      @rel = options[:rel]
+      @target = options[:target]
+      @template = options[:template]
+      @curie_resolver = options[:curie_resolver] || CurieResolver.new([])
+
+      (fail ArgumentError, "A rel must be provided") if @rel.nil?
+
+      if @target.nil? && @template.nil?
+        (fail ArgumentError, "A target or template must be provided")
+      end
+
+      if @target && @template
+        (fail ArgumentError, "Cannot provide both a target and a template")
+      end
+
+      if @target && !@target.kind_of?(Representation)
+        (fail ArgumentError, "Invalid HAL representation: #{target.inspect}")
+      end
+
+      if @template && !@template.kind_of?(Addressable::Template)
+        (fail ArgumentError, "Invalid Addressable::Template: #{template.inspect}")
+      end
+    end
+
+    attr_accessor :rel, :target, :template, :curie_resolver
+
+    def self.new_from_link_entry(hash_entry:, hal_client:, curie_resolver:, base_url:)
+      rel = hash_entry[:rel]
+      hash_data = hash_entry[:data]
+      href = (Addressable::URI.parse(base_url) + hash_data['href']).to_s
+
+      if hash_data['templated']
+        Link.new(rel: rel,
+                 template: Addressable::Template.new(href),
+                 curie_resolver: curie_resolver)
+      else
+        Link.new(rel: rel,
+                 target: Representation.new(hal_client: hal_client, href: href),
+                 curie_resolver: curie_resolver)
+      end
+    end
+
+    def self.new_from_embedded_entry(hash_entry:, hal_client:, curie_resolver:, base_url:)
+      rel = hash_entry[:rel]
+      hash_data = hash_entry[:data]
+
+      absolute_href = (Addressable::URI.parse(base_url) + hash_data['_links']['self']['href']).to_s
+      hash_data['_links']['self']['href'] = absolute_href
+
+      Link.new(rel: rel,
+               target: Representation.new(hal_client: hal_client, parsed_json: hash_data),
+               curie_resolver: curie_resolver)
+    end
+
+
+    # Returns the URL of the resource this link references.
+    # In the case of a templated link, this is the unresolved url template pattern.
+    def raw_href
+      templated? ? template.pattern : target.href
+    end
+
+    def resolved_rel
```

Comment:
I think `fully_qualified_rel` would be a better name. And maybe `literal_rel` for the other version?  `rel` alone is a little ambiguous so maybe we should just not use it?